src/models/room.rs

```diff
@@ -0,0 +1,298 @@
+// Copyright 2020 Damir Jelić
+// Copyright 2020 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::sync::{Arc, RwLock};
+
+use super::{RoomMember, User, UserId};
+
+use crate::events::collections::all::{RoomEvent, StateEvent};
+use crate::events::room::{
+    aliases::AliasesEvent,
+    canonical_alias::CanonicalAliasEvent,
+    member::{MemberEvent, MembershipChange},
+    name::NameEvent,
+    power_levels::PowerLevelsEvent,
+};
+use crate::events::{
+    presence::{PresenceEvent, PresenceEventContent},
+    EventResult,
+};
+use crate::identifiers::RoomAliasId;
+use crate::session::Session;
+
+use js_int::UInt;
+
+#[cfg(feature = "encryption")]
+use tokio::sync::Mutex;
+
+#[cfg(feature = "encryption")]
+use crate::crypto::{OlmMachine, OneTimeKeys};
+#[cfg(feature = "encryption")]
+use ruma_client_api::r0::keys::{upload_keys::Response as KeysUploadResponse, DeviceKeys};
+
+#[derive(Debug, Default)]
+/// `RoomName` allows the calculation of a text room name.
+pub struct RoomName {
+    /// The displayed name of the room.
+    name: Option<String>,
+    /// The canonical alias of the room ex. `#room-name:example.com` and port number.
+    canonical_alias: Option<RoomAliasId>,
+    /// List of `RoomAliasId`s the room has been given.
+    aliases: Vec<RoomAliasId>,
+}
+
+#[derive(Debug)]
+/// A Matrix rooom.
+pub struct Room {
+    /// The unique id of the room.
+    pub room_id: String,
+    /// The name of the room, clients use this to represent a room.
+    pub room_name: RoomName,
+    /// The mxid of our own user.
+    pub own_user_id: UserId,
+    /// The mxid of the room creator.
+    pub creator: Option<UserId>,
+    /// The map of room members.
+    pub members: HashMap<UserId, RoomMember>,
+    /// A list of users that are currently typing.
+    pub typing_users: Vec<UserId>,
+    /// A flag indicating if the room is encrypted.
+    pub encrypted: bool,
```

Comment:
This can be a bool for now (I know I wrote this) but we'll need to store the algorithm and rotation time we get from the event that marks the room as encrypted. If you add handling for that event type in the future.